configtx/application.go

```diff
@@ -423,15 +431,359 @@ func (a *ApplicationOrg) setMSPConfig(updatedMSP MSP) error {
+	if err != nil {
+		return err
+	}
+
+	for _, c := range msp.Admins {
+		if c.Equal(cert) {
+			return nil
+		}
+	}
+
+	msp.Admins = append(msp.Admins, cert)
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// RemoveAdminCert removes an administator identity from the organization MSP.
+func (m *ApplicationOrgMSP) RemoveAdminCert(cert *x509.Certificate) error {
+	msp, err := getMSPConfig(m.configGroup)
+	if err != nil {
+		return err
+	}
+
+	certs := msp.Admins[:]
+	for i, c := range msp.Admins {
+		if c.Equal(cert) {
+			certs = append(certs[:i], certs[i+1:]...)
+			break
+		}
+	}
+
+	msp.Admins = certs
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// AddRootCert adds an root certificate trusted by the organization MSP.
+func (m *ApplicationOrgMSP) AddRootCert(cert *x509.Certificate) error {
+	msp, err := getMSPConfig(m.configGroup)
+	if err != nil {
+		return err
+	}
+
+	for _, c := range msp.RootCerts {
+		if c.Equal(cert) {
+			return nil
+		}
+	}
+
+	msp.RootCerts = append(msp.RootCerts, cert)
+
+	err = msp.validateCACerts()
+	if err != nil {
+		return err
+	}
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// RemoveRootCert removes an trusted root certificate from the organization MSP.
+func (m *ApplicationOrgMSP) RemoveRootCert(cert *x509.Certificate) error {
+	msp, err := getMSPConfig(m.configGroup)
 	if err != nil {
-		return nil, fmt.Errorf("retrieving application org msp: %s", err)
+		return err
+	}
+
+	certs := msp.RootCerts[:]
+	for i, c := range msp.RootCerts {
+		if c.Equal(cert) {
+			certs = append(certs[:i], certs[i+1:]...)
+			break
+		}
+	}
+
+	msp.RootCerts = certs
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// AddIntermediateCert adds an intermediate certificate trusted by the organization MSP.
+func (m *ApplicationOrgMSP) AddIntermediateCert(cert *x509.Certificate) error {
+	msp, err := getMSPConfig(m.configGroup)
+	if err != nil {
+		return err
+	}
+
+	for _, c := range msp.IntermediateCerts {
+		if c.Equal(cert) {
+			return nil
+		}
+	}
+
+	msp.IntermediateCerts = append(msp.IntermediateCerts, cert)
+
+	err = msp.validateCACerts()
+	if err != nil {
+		return err
+	}
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// RemoveIntermediateCert removes an trusted intermediate certificate trusted by the organization MSP.
+func (m *ApplicationOrgMSP) RemoveIntermediateCert(cert *x509.Certificate) error {
+	msp, err := getMSPConfig(m.configGroup)
+	if err != nil {
+		return err
+	}
+
+	certs := msp.IntermediateCerts[:]
+	for i, c := range msp.IntermediateCerts {
+		if c.Equal(cert) {
+			certs = append(certs[:i], certs[i+1:]...)
+			break
+		}
+	}
+
+	msp.IntermediateCerts = certs
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// AddOUIdentifier adds a fabric organizational unit for the organization MSP.
+func (m *ApplicationOrgMSP) AddOUIdentifier(ou membership.OUIdentifier) error {
+	msp, err := getMSPConfig(m.configGroup)
+	if err != nil {
+		return err
+	}
+
+	for _, o := range msp.OrganizationalUnitIdentifiers {
```

Comment:
We may want to consider exactly what the behavior should be when adding duplicates for the name. A `membership.OUIdentifier` contains both a certificate as well as the string identifier. Effectively it can almost be treated as a map from identifier to the CA under which identities with that OU are validated, although for purposes of marshaling we treat it as a slice of a type that contains both the certificate and identifier. You may want to sync with MSP experts on whether an MSP can be configured with multiple certificates for the same OUIdentifier string. If that's not supported we may want to add an error here or alternatively change the functionality here to `SetOUIdentifier` where if the identifier string in the `membership.OUIdentifier` is already found in the existing config, it would just replace the certificate with the new one passed in since otherwise I see no path for modifying the existing OU identifiers without removing the old one to add a new one with the same string (effectively a replace).

Reply:
Are you okay with investigating this separately?